subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -513,47 +520,76 @@ int fota_download_start_with_image_type(const char *host, const char *file,
 	return 0;
 }
 
-int fota_download_init(fota_download_callback_t client_callback)
+static int fota_download_object_init(void)
+{
+	int err;
+
+#ifdef CONFIG_FOTA_DOWNLOAD_NATIVE_TLS
+	/* Enable native TLS for the download client socket
+	 * if configured.
+	 */
+	dlc.set_native_tls = CONFIG_FOTA_DOWNLOAD_NATIVE_TLS;
+#endif
+
+	k_work_init_delayable(&dlc_with_offset_work, download_with_offset);
+
+	err = download_client_init(&dlc, download_client_callback);
+	if (err != 0) {
+		return err;
+	}
+
+	initialized = true;
+	return 0;
+}
+
+int fota_download_client_init(fota_download_callback_t client_callback, bool smp_image_type)
 {
 	int err;
-	static bool initialized;
 
 	if (client_callback == NULL) {
 		return -EINVAL;
 	}
 
 	callback = client_callback;
+	err = fota_download_util_stream_init();
+	if (err) {
+		LOG_ERR("%s failed to init stream %d", __func__, err);
+		return err;
+	}
+	use_smp_dfu_target = smp_image_type;
 
 	if (initialized) {
 		return 0;
 	}
 
-#ifdef CONFIG_FOTA_DOWNLOAD_NATIVE_TLS
-	/* Enable native TLS for the download client socket
-	 * if configured.
-	 */
-	dlc.set_native_tls = CONFIG_FOTA_DOWNLOAD_NATIVE_TLS;
+	return fota_download_object_init();
+}
+
+int fota_download_init(fota_download_callback_t client_callback)
+{
+#ifdef CONFIG_DFU_TARGET_MCUBOOT
+	int err;
```

Comment:
nit: would it be better to declare this in the already needed #ifdef block further down? So at line 580 in the current state.